Allow attaching File or Pathname to has_one_attached

dorianmariecom opened the pull request:
### Why was this change necessary?

When creating models in tests, it's easier to pass a File or a Pathname
(from `file_fixture` for instance) to `Model.create` for instance

### How does it address the problem?

When attaching an attachable, we check if it's a File or a Pathname and
handle it appropriately.

seanpdoyle commented on activestorage/lib/active_storage/attached/changes/create_one.rb:
```diff
@@ -70,6 +70,28 @@ def find_or_build_blob
+          )
+        when Pathname
+          filename = File.basename(attachable.basename)
+          content_type = Marcel::MimeType.for(attachable.open, name: filename)
```
Does invoking `attachable.open` twice (once here, and once in the `io:` keyword argument below) pose any issues here?

tenderlove replied:
We should probably not call `open` twice as it'll use two file descriptors.

dorianmariefr replied:
I've removed the second call to open

seanpdoyle commented on activestorage/lib/active_storage/attached/changes/create_one.rb:
```diff
@@ -70,6 +70,28 @@ def find_or_build_blob
           )
         when String
           ActiveStorage::Blob.find_signed!(attachable, record: record)
+        when File
```
The `File` and `Pathname` branches have almost identical structures. Would it be worth trying to combine them?

```ruby
when File, Pathname
  io = attachable.is_a?(File) ? attachable : attachable.open
  filename = File.basename(attachable.path)
  content_type = Marcel::MimeType.for(io, name: filename)

  ActiveStorage::Blob.create_and_upload!(
    io: io,
    filename: filename,
    content_type: content_type,
    record: record,
    service_name: attachment_service_name
  )
end
```

tenderlove replied:
I think this would be a good refactoring!

dorianmariefr replied:
I've applied the refactoring and ran the tests and they pass

Bitcoven commented:
I am very happy very very good

TylerRick commented:
Does this just need to be rebased so it can be merged?

dorianmariefr commented:
@tenderlove could you take another look plz?

sebroeder commented:
It would be amazing if this PR did land in main before Rails 7.1 🤞🙏. Let me know if I can help in any way.

p8 commented on activestorage/test/models/attached/one_test.rb:
```diff
@@ -16,6 +16,22 @@ class ActiveStorage::OneAttachedTest < ActiveSupport::TestCase
     ActiveStorage::Blob.all.each(&:delete)
   end
 
+  test "creating an user with a File as avatar" do
```
Tiny nitpick, the first sound of user is not a vowel so it should be:
```suggestion
  test "creating a user with a File as avatar" do
```
https://english.stackexchange.com/questions/105116/is-it-a-user-or-an-user

p8 replied:
As the tests shouldn't care about the type of record or the attachment name, what do you think about:
```suggestion
  test "creating a record with a File as attached" do
```

p8 commented on activestorage/test/models/attached/one_test.rb:
```diff
@@ -16,6 +16,22 @@ class ActiveStorage::OneAttachedTest < ActiveSupport::TestCase
+    assert_not_nil @user.avatar_blob
+  end
+
+  test "creating an user with a Pathname as avatar" do
```
Same...
```suggestion
  test "creating a user with a Pathname as avatar" do
```

p8 replied:
```suggestion
  test "creating a record with a Pathname as attached" do
```

p8 commented on activestorage/test/fixtures/files/dorian.jpg:
Ha nice 😄 . You'll be part of the Rails repo if this gets merged.

dorianmariefr replied:
Yeah, I thought it would be cool

seanpdoyle commented on activestorage/test/models/attached/one_test.rb:
```diff
@@ -16,6 +16,22 @@ class ActiveStorage::OneAttachedTest < ActiveSupport::TestCase
   end
 
+  test "creating a record with a File as attachable attribute" do
+    @user = User.create!(name: "Dorian", avatar: File.open("test/fixtures/files/dorian.jpg"))
```
Since files are defined in `test/fixtures/files`, would a call to `file_fixture(...).open` return the same `File` value as `File.open` with the full path?

dorianmariefr replied:
yeah that seems to do the same thing:

```
>> File.open("test/fixtures/files/dorian.jpg")
=> #<File:test/fixtures/files/dorian.jpg>
>> file_fixture("dorian.jpg").open
=> #<File:/Users/dorianmariefr/src/rails/activestorage/test/fixtures/files/dorian.jpg>
```

p8 commented on activestorage/lib/active_storage/attached/changes/create_one.rb:
```diff
@@ -82,6 +82,18 @@ def find_or_build_blob
+          filename = File.basename(attachable)
+          content_type = Marcel::MimeType.for(io, name: filename)
+
+          ActiveStorage::Blob.create_and_upload!(
+            io: io,
+            filename: filename,
+            content_type: content_type,
+            record: record,
+            service_name: attachment_service_name
+          )
```
`create_and_upload!` will try to identify the `content_type` by default.
We can pass `identify: false`, or let ActiveStorage identify the file:
https://github.com/rails/rails/blob/b1ac22fe7edffaf47051c207d2c43d8971b98a25/activestorage/app/models/active_storage/blob.rb#L97-L101

```suggestion
          ActiveStorage::Blob.create_and_upload!(
            io: io,
            filename: filename,
            content_type: content_type,
            record: record,
            identify: false,
            service_name: attachment_service_name
          )
```

p8 replied:
Sorry, that is not correct, the documentation is incorrect.
If content_type is set it won't identity.

chaadow commented on activestorage/lib/active_storage/attached/changes/create_one.rb:
```diff
@@ -82,6 +82,18 @@ def find_or_build_blob
+          filename = File.basename(attachable)
+          content_type = Marcel::MimeType.for(io, name: filename)
+
+          ActiveStorage::Blob.create_and_upload!(
```
Maybe use `build_after_unfurling` instead of `create_and_upload!` like the other branches. Plus, the method's name is `find_or_build_blob`

dorianmariefr replied:
Good call, I don't know why I was using `create_and_upload!`, `build_after_unfurling` works well

dorianmariecom committed:
Allow attaching File or Pathname to has_one_attached

### Why was this change necessary?

When creating models in tests, it's easier to pass a File or a Pathname
(from `file_fixture` for instance) to `Model.create` for instance

### How does it address the problem?

When attaching an attachable, we check if it's a File or a Pathname and
handle it appropriately.

p8 commented:
With the change to use `build_after_unfurling` does the following work as well?
```ruby
User.new(avatar: File.open("image.jpg"))
```

dorianmariefr commented:
@p8 File.open(...) and file_fixture(...).open both return a File so they both work

byroot reviewed (approved):
Just one minor thing, other than that LGTM.

dorianmariefr commented on activestorage/test/models/attached/one_test.rb:
```diff
@@ -16,6 +16,22 @@ class ActiveStorage::OneAttachedTest < ActiveSupport::TestCase
+  test "creating a record with a File as attachable attribute" do
+    @user = User.create!(name: "Dorian", avatar: file_fixture("dorian.jpg").open)
+
+    assert_equal "dorian.jpg", @user.avatar.filename.to_s
```
done 😭

dorianmariefr commented:
It's weird, it's supposed to be saved locally but it's not saving: (the me/ is a `attach(io: ..., filename: ...)` test)

```
>> ActiveStorage::Blob.service.path_for(User.last.avatar.key)
=> "/Users/dorianmariefr/src/factory-bot-has-one-attached/storage/ej/i3/eji3y47ftozu14iomicdx0n5nbq0"
```

```
> ls storage/
me/
```

dorianmariefr commented:
Found why, it's not handling the upload, will make a follow-up pull request https://github.com/rails/rails/blob/main/activestorage/lib/active_storage/attached/changes/create_one.rb#L23-L34

dorianmariefr commented:
Follow-up pull request that fixes the uploading of the files https://github.com/rails/rails/pull/49381

Bitcoven commented:
***@***.***

On Mon, Sep 25, 2023, 22:19 Dorian Marié (perso) ***@***.***>
wrote:

> Follow-up pull request that fixes the uploading of the files #49381
> <https://github.com/rails/rails/pull/49381>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/45606#issuecomment-1734165688>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AW64F2WHLNQDXMWPXLTFBVDX4G4LLANCNFSM53VOJDCQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
